cdap-watchdog/src/main/java/co/cask/cdap/metrics/data/TimeSeriesTable.java

```diff
@@ -148,6 +150,147 @@ public MetricsScanner scan(MetricsScanQuery query) throws OperationException {
+    int endTimeBase = getTimeBase(query.getEndTime());
+    byte[][] columns = null;
+    String tagPrefix = query.getTagPrefix();
+    if (tagPrefix == null) {
+      tagPrefix = MetricsConstants.EMPTY_TAG;
+    }
+
+    byte[] startRow = entityCodec.paddedEncode(query.getContextPrefix(), query.getMetricPrefix(), tagPrefix,
+                                        startTimeBase, query.getRunId(), 0);
+    byte[] endRow = entityCodec.paddedEncode(query.getContextPrefix(), query.getMetricPrefix(), tagPrefix,
+                                      endTimeBase + 1, query.getRunId(), 0xff);
+
+    Row rowResult;
+    // does multiple scan, but we skip duplicate row-keys by incrementing
+    // the sub-context or metrics part of rowkey before each scan.
+    do {
+      FuzzyRowFilter filter = getFilter(query, startTimeBase, endTimeBase + 1, false, FOUR_ONE_BYTES);
+
+      ScannerFields fields = new ScannerFields(startRow, endRow, columns, filter);
+      Scanner scanner = null;
+      try {
+        scanner = timeSeriesTable.scan(fields.startRow, fields.endRow, fields.columns, fields.filter);
+      } catch (Exception e) {
+        throw new OperationException(StatusCode.INTERNAL_ERROR, e.getMessage(), e);
+      }
+
+      rowResult = scanner.next();
+      if (rowResult != null) {
+        byte[] rowKey = rowResult.getRow();
+        // Decode context and metric from key
+        int offset = 0;
+        String contextStr = entityCodec.decode(MetricsEntityType.CONTEXT, rowKey, offset);
+
+        // Always have a "." suffix for unique matching
+        if (query.getContextPrefix() != null && !(contextStr + ".").startsWith(query.getContextPrefix())) {
+          scanner.close();
+          break;
+        }
+
+        offset += entityCodec.getEncodedSize(MetricsEntityType.CONTEXT);
+        String metricName = entityCodec.decode(MetricsEntityType.METRIC, rowKey, offset);
+
+        if (isContextQuery) {
+          // update the rowkey, by incrementing the next-level context part
+          ByteBuffer contextPart = ByteBuffer.wrap(rowKey);
+          contextPart.limit(entityCodec.getEncodedSize(MetricsEntityType.CONTEXT));
+          contextPart = getNextContextTarget(contextPart, query.getContextPrefix());
+          if (contextPart == null) {
+            scanner.close();
+            break;
+          }
+          //update the start-row to be row-key for next scan
+          startRow = rowKey;
+          metricsScanResults.add(contextStr);
+        } else {
+          // update the rowkey, by incrementing the metrics part
+          ByteBuffer metricsPart = ByteBuffer.wrap(rowKey);
+          metricsPart.position(entityCodec.getEncodedSize(MetricsEntityType.CONTEXT));
+          metricsPart.limit(entityCodec.getEncodedSize(MetricsEntityType.CONTEXT) +
+                              entityCodec.getEncodedSize(MetricsEntityType.METRIC));
+          metricsPart = stopKeyForPrefix(metricsPart);
```

Comment:
```
// With the next scan we fast-forward to the next row that has different context or metric name part 
// that we are searching for.
startRow = getNextRow(rowKey, offset, length);

// Gets row key that is:
// * greater than given rowKey AND
// * has same bytes in the beginning (up to the given offset) AND
// * has different bytes in the range [offset..offset+length]
// or null if no row key with this properties exist.
byte[] getNextRow(rowKey, offset, length) {...}
```
in implementation do any simplest possible, e.g. use arrayCopy and Bytes.stopKeyForPrefix().